README.md

```diff
@@ -481,61 +448,105 @@ Match a sequence of expressions and return their match results in an array.
 etc. Return the match result of the first successfully matched expression. If no
 expression matches, consider the match failed.
 
+### Action Execution Environment
+
+Actions and predicates have these variables and functions
+available to them.
+
+- All variables and functions defined in the initializer or the top-level
+  initializer at the beginning of the grammar are available.
+
+- Note, that all functions and variables, described below, is unavailable
+  in the global initializer.
+
+- Labels from preceding expressions are available as local variables,
+  which will have the match result of the labelled expressions.
+
+  A label is only available after its labelled expression is matched:
+
+  ```pegjs
+  rule = A:('a' B:'b' { /* B is available, A is not */ } )
+  ```
+
+  A label in a sub-expression is only valid within the
+  sub-expression:
+
+  ```pegjs
+  rule = A:'a' (B: 'b') (C: 'b' { /* A and C are available, B is not */ })
+  ```
+
+- `input` is an parsed string that was passed to the `parse()` method.
+
+- `options` is a variable that contains the parser options.
+  That is the same object that was passed to the `parse()` method.
+
+- `error(message, where)` will report an error and throw an exception.
+  `where` is optional; the default is the value of `location()`.
+
+- `expected(message, where)` is similar to `error`, but reports
+  > Expected _message_ but "_other_" found.
+
+- `location()` returns an object like this:
+
+  ```javascript
+  {
+    source: options.grammarSource,
+    start: { offset: 23, line: 5, column: 6 },
+    end: { offset: 25, line: 5, column: 8 }
+  }
+  ```
+
+  `source` is an any object that was suplied in the `grammarSource` option in
+  the `parse` call. That object can be used to hold reference to the origin of
+  the grammar, for example, it can be a filename. It is recommended that this
+  object have a `toString()` implementation that returns meaningful string,
+  because that string will be used when getting formatted error representation
+  with `e.format()`.
+
+  If `source` is `null` or `undefined` it don't appeared in the formatted messages.
+  Default value for `source` is `undefined`.
+
+  For actions, `start` refers to the position at the beginning of the preceding
+  expression, and `end` refers to the position after the end of the preceding
+  expression.
+
+  For predicates, `start` and `end` are the same, the location where the predicate
+  is evaluated.
+
+  For initializer, the location is the start location, i.e.
+  ```javascript
+  {
+    source: options.grammarSource,
+    start: { offset: 0, line: 1, column: 1 },
+    end: { offset: 0, line: 1, column: 1 }
+  }
+  ```
+
+  `offset` is a 0-based character index within the source text.
+  `line` and `column` are 1-based indices.
+
+- `text()` returns the source text between `start` and `end` (which will be `""` for
+  predicates). Instead of using that function as a return value for the rule consider
```

Comment:
predicates not having access to `text()` feels like a bug.  Opinions?

Reply:
No, it not a bug: predicates points just to the single position, so what would `text()` returns for them? `text()` just takes part of input on which action expression is matched. Semantic predicates even does not "match" anything, so you shouldn't be surprised that `text()` returns the empty string

Reply:
That makes sense, thanks for the explanation.